src/macroxs.F90

```diff
@@ -0,0 +1,243 @@
+module macroxs
+
+  use constants
+  use macroxs_header,  only: MacroXS_Base, MacroXS_Iso, MacroXS_Angle, &
+                             expand_harmonic
+  use material_header, only: Material
+  use math
+  use nuclide_header,  only: find_angle, MaterialMacroXS, NuclideMicroXS, &
+                             Nuclide_MG, NuclideMGContainer
+  use random_lcg,      only: prn
+  use scattdata_header
+  use search
+
+  implicit none
+
+contains
+
+!===============================================================================
+! UPDATE_XS stores the xs to work with
+!===============================================================================
+
+  subroutine calculate_mgxs(this, gin, uvw, xs)
+    class(MacroXS_Base),   intent(in)    :: this
+    integer,               intent(in)    :: gin         ! Incoming neutron group
+    real(8),               intent(in)    :: uvw(3)      ! Incoming neutron direction
+    type(MaterialMacroXS), intent(inout) :: xs
```

Comment:
The standard OpenMC way of doing this would have been to take only the particle `p` as an input argument and then modify the global `material_xs`.

I like that what you have is more explicit, but I think I'd slightly prefer you switch to `p` argument and global `material_xs` so that it more closely matches the `calculate_xs` subroutine.